pykern/fconf.py

```diff
@@ -0,0 +1,569 @@
+    YAML files next. YAML file evaluation happens in that same order.
+
+    Args:
+        path (py.path): directory that *.py and *.yml files
+
+    """
+    return Parser(
+        pykern.pkio.sorted_glob(path.join("*.py"))
+        + pykern.pkio.sorted_glob(path.join("*.yml")),
+    ).result
+
+
+class Parser(PKDict):
+    def __init__(self, files):
+        self.pkupdate(
+            files=PKDict(),
+            macros=PKDict(),
+        )
+        self._add_file(pykern.pkio.py_path(f"fconf.{_BUILTINS_EXT}"))
+        for f in files:
+            try:
+                self._add_file(f)
+            except Exception as e:
+                _exception_reason(e, f"fconf.Parser.file={f}")
+                raise
+        e = _Evaluator(parser=self)
+        for f in self.files.yml:
+            e.start(source=f)
+        self.result = e.global_fvars
+
+    def _add_file(self, path):
+        e = path.ext[1:].lower()
+        f = _File(
+            content=getattr(self, f"_ext_{e}")(path),
+            ext=e,
+            path=path,
+        )
+        if e in ("py", _BUILTINS_EXT):
+            self._add_macros(f)
+        elif e == "yml":
+            self._add_text_macros(f)
+        else:
+            raise ValueError(f"unhandled file ext={e}")
+        self.files.setdefault(e, []).append(f)
+
+    def _add_macro(self, macro):
+        n = macro.name
+        if n in self.macros:
+            raise ValueError(f"duplicate {macro}, other {self.macros[macro.name]}")
+        if n.startswith(_RESERVED_PREFIX) and macro.source.ext != _BUILTINS_EXT:
+            raise ValueError(f"macro={macro} may not begin with {_RESERVED_PREFIX}")
+        self.macros[macro.name] = macro
+
+    def _add_macros(self, source):
+        m = source.content
+        if not m:
+            return
+        for n, f in m.items():
+            self._add_macro(
+                _Macro(
+                    func=f,
+                    name=n,
+                    # TODO(robnagler) assert varnames don't begin with fconf_
+                    params=tuple(f.__code__.co_varnames),
+                    source=source,
+                ),
+            )
+
+    def _add_text_macros(self, source):
+        m = source.content.pkdel("fconf_macros")
+        if not m:
+            return
+        for n, c in m.items():
+            m = _TEXT_MACRO_DEF.search(n)
+            if not m:
+                raise ValueError(f"invalid macro name={n} {source}")
+            n = m.group(1)
+            # TODO: duplicate arg check
+            self._add_macro(
+                _YAMLMacro(
+                    content=c,
+                    name=n,
+                    # TODO(robnagler): check for dups
+                    # TODO(robnagler): assert varnames don't begin with fconf_
+                    params=tuple((x for x in _ARG_SEP.split(m.group(2)) if x)),
+                    source=source,
+                ),
+            )
+
+    def _ext_builtins(self, path):
```

Comment:
path isn't used

Reply:
Yes, and it's called by a dispatch so path is always passed.